cwms-data-api/src/main/java/cwms/cda/formatters/json/adapters/TimeSeriesRecordDeserializer.java

```diff
@@ -0,0 +1,77 @@
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ *  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package cwms.cda.formatters.json.adapters;
+
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.databind.DeserializationContext;
+import com.fasterxml.jackson.databind.JsonDeserializer;
+import com.fasterxml.jackson.databind.JsonNode;
+import cwms.cda.data.dto.TimeSeries;
+import cwms.cda.data.dto.TimeSeriesRecordWithDate;
+import java.io.IOException;
+import java.sql.Timestamp;
+import java.time.Instant;
+
+/**
+ * A time-series record deserializer for properly deserializing JSON data.
+ * Requires {@link cwms.cda.data.dto.TimeSeries.RecordChild} class to avoid
+ * a StackOverflowError when deserializing JSON data. This issue can be caused by the custom serializer
+ * getting stuck in a loop if the Record class is used directly.
+ * Allows for use of subclass with additional fields {@link TimeSeriesRecordWithDate}.
+ */
+public final class TimeSeriesRecordDeserializer extends JsonDeserializer<TimeSeries.Record> {
+    @Override
+    public TimeSeries.Record deserialize(JsonParser jsonParser, DeserializationContext deserializationContext) throws IOException {
+        JsonNode node = jsonParser.readValueAsTree();
+        if (node.get("data-entry-date") != null) {
+            return jsonParser.getCodec().treeToValue(node, TimeSeriesRecordWithDate.class);
+        }
+        String nodeString = node.toString();
+        if (nodeString.startsWith("[")) {
+            nodeString = nodeString.substring(1, nodeString.length() - 1);
+            String[] valList = nodeString.split(",");
+            if (valList.length == 4) {
+                Timestamp dateTime = Timestamp.from(Instant.ofEpochMilli(Long.parseLong(valList[0])));
+                double value = Double.parseDouble(valList[1]);
+                int quality = Integer.parseInt(valList[2]);
+                Timestamp entryDate = Timestamp.from(Instant.ofEpochMilli(Long.parseLong(valList[3])));
```

Comment:
no need to convert from Instant to Timestamp, Timestamp's constructor takes in epoch millis.